designs/odcr.md

```diff
@@ -0,0 +1,121 @@
+- Supporting changes in scaling behavior when ODCR is associated to a NodeClass. _We won't bring up N nodes to match an N node capacity reservation_
+
+## Proposed Solution
+
+### Supporting associating ODCR to EC2NodeClass
+
+Add a new field `capacityReservationSpec` to `EC2NodeClass` 
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: example-node-class
+spec:
+  capacityReservationSpec:
+    capacityReservationPreference: open | none | None  # Cannot be defined if capacityReservationTarget is specified
+    capacityReservationTarget: # Cannot be defined if capacityReservationPreference is specified
+      capacityReservationId: String | None
+      capacityReservationResourceGroupArn: String | None
+```
+This follows exactly how LaunchTemplate defines [CapacityReservationSpecification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-capacityreservationspecification.html) with the same
+constraints. 
+
+Karpenter will perform validation against the spec to ensure there isn't any violation prior to creating the LaunchTemplates.
+
+### Launching Nodes into Capacity Reservation
+
+#### All Launch Templates are associated with the specified Capacity Reservation
+
+EC2NodeClass currently supports automatically generating Launch Templates based on instance types and their AMI requirements however as the first iteration of supporting Capacity Reservation, we won't introduce any prioritization nor fallback nor safeguard when launching nodes against EC2NodeClasses with a Capacity Reservation. This means that it is the
+user's responsibility to ensure the Capacity Reservation associated to a EC2NodeClasses will include instance types the EC2NodeClasses could launch. Otherwise the EC2NodeClasses will fail during provisioning.
+
+This also means that we will not allow a EC2NodeClass be able to create Spot instances when user specified a Capacity Reservation. 
+
+We will skip [checkODFallback](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instance/instance.go#L200C14-L200C29).
+
+_Note that these aren't permanent restrictions but simply narrowing down what features exist in the first iteration of supporting Capacity Reservation_ 
+
+Pros: 
+- This puts the onus of checking on the user. It doesn't require checking if a Capacity Reservation satisfy the requirements of an instance type by assuming the users is doing this. It makes no assumption about what a Capacity Reservation can do nor implies can do. This is helpful because if there is expectation that Karpenter does this we may be asked to
+support checking if there is capacity in the reservation before launching nodes which would significantly increase the scope of this design.
+
+Cons:
+- This implementation is overly restrictive causing it to be difficult to use. Its possible that the restrictions makes it too hard for users to use EC2NodeClasses effectively.
+
+Both a pro and a con:
```

Comment:
In general, I think this is a good thing. When it comes to fallback and prioritization, I think this makes sense. You have a set of instance types that you prioritize for usage with the ODCR and then you can create a lower-weighted NodePool that has a broader set of instance types or perhaps uses the same instance type that you had for the ODCR but with spot and on-demand to try to get that type again just like from a normal request

Reply:
Agreed